src/gravel/controllers/kv.py

```diff
@@ -12,125 +12,395 @@
+                        self._cluster.create_pool("aquarium")
+                    self._ioctx = self._cluster.open_ioctx("aquarium")
+                    # This actually seems to be safe (doesn't trash existing omap
+                    # data if present, which is neat)
+                    self._ioctx.write_full(
+                        "kvstore",
+                        "# aquarium kv store is in this object's omap\n".encode(
+                            "utf-8"
+                        ),
+                    )
+                    # At this point, iff it's a new pool, new object, etc.
+                    # we need to push everything from our local cache to
+                    # the omap on our kvstore, to populate it with whatever
+                    # may have been set pre-bootstrap.
+                    # Probably want to lock the object here too, in case we get two
+                    # simultaenous starts (and thus omap population)
+                    keys = self._db.keys()
+                    values = list(self._db[k] for k in keys)
+                    if keys and not has_aquarium_pool:
+                        with rados.WriteOpCtx() as op:
+                            logger.info(
+                                f"Pushing {keys} to newly created aquarium pool"
+                            )
+                            self._ioctx.set_omap(op, keys, values)  # type: ignore
+                            self._ioctx.operate_write_op(op, "kvstore")
+                    # TODO: need timeouts, error handlers etc on watch
+                    self._config_watch = self._ioctx.watch(
+                        "kvstore", self._config_notify
+                    )
+                    logger.debug(
+                        f"config watch id is {self._config_watch.get_id()}"
+                    )
+                    # will raise:
+                    # rados.ObjectNotFound: [errno 2] RADOS object not found (watch error)
+            except Exception as e:
+                # TODO: deal with this (log it? ignore it?)
+                # e.g. RADOS rados state (You cannot perform that operation on a Rados object in state configuring.)
+                # this makes the log pretty messy prior to bootstrap (errors every 10 seconds)
+                logger.error(str(e))
+
+            # TODO: Should we sleep for longer?  A minute instead of 10 seconds?  This is pretty arbitrary...
+            logger.debug("Cluster connection thread sleeping for 10 seconds")
+            self._event.wait(10)
+            self._event.clear()
+            # How on earth to we detect that the cluster has gone away
+            # and trigger a reconnect?
+            # What happens if we try a write op on a down cluster?
+            # might want watch checks?
+
+        logger.debug("Shutting down cluster connection")
+        if self._config_watch:
+            self._config_watch.close()
+            # Need to set this to None, so it's deallocated before the
```

Comment:
This might be a python librados bug where `.close()` should actually clean things up (but is reasonable for here)

Reply:
I think I'm using the watch in a way that isn't usually tested, i.e. it's not in a `with` block, and I'm actually explicitly calling `Rados.shutdown()` elsewhere. The offending `__dealloc__` code is https://github.com/ceph/ceph/blob/a9a4cc9abf201eca9c11a0bf82eedd14f2343fbb/src/pybind/rados/rados.pyx#L2027-L2029. The fix in rados (to avoid having to set the object to `None` after calling `Watch.close()`) is probably this:

```diff
--- a/src/pybind/rados/rados.pyx
+++ b/src/pybind/rados/rados.pyx
@@ -2025,6 +2025,8 @@ cdef class Watch(object):
         return False
 
     def __dealloc__(self):
+        if self.id == 0:
+            return
         self.ioctx.rados.require_state("connected")
         self.close()
 ```

Of course, now that I think of it, it would be enough for me to set the object to None, and _not_ call close, then I'd get the automatic close-on-dealloc. But, I _still_ think it's a bug in librados that an already-explicitly-closed watch raises an error when it goes out of scope after the cluster shuts down, so I've made a note to do something about addressing that upstream :-)

Reply:
...and here's the proposed librados fix: https://github.com/ceph/ceph/pull/43107